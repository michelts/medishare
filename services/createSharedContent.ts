import { v4 as uuidv4 } from 'uuid';
import axios from 'axios';

type SharedContent = {
  id: string,
  name: string,
  filename: string
};

export default async function createSharedContent({ name, file }: { name: string, file: string }): SharedContent {
  const id = uuidv4()
  const { url, fields, filename } = await getSignedPostUrl({ id });
  await postFile({ url, fields, file });
  return await saveSharedContent({ id, name, filename });
}

async function getSignedPostUrl({ id }) {
  const { data: { url, fields, filename } } = await axios.post('/api/v1/upload-urls/', { id });
  return { url, fields, filename };
}

async function postFile({ url, fields, file }) {
  const payload = new FormData()
  Object.keys(fields).forEach(key => payload.append(key, fields[key]))
  payload.append('file', file)
  await axios.post(url, payload);
}

async function saveSharedContent({ id, name, filename }) {
  const { data } = await axios.post('/api/v1/shared-contents/', { id, name, filename })
  return data;
}
