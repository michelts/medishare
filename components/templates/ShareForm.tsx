import { useState, useEffect, useCallback } from 'react';
import ReactVideoRecorder from 'react-video-recorder';
import Link from "next/link"
import createSharedContent from '@services/createSharedContent';
import Anchor from '@atoms/Anchor';
import Loading from '@atoms/Loading';
import CopyToClipboard from '@atoms/CopyToClipboard';

const ShareForm: React.FC = () => {
  const [mode, setMode] = useState(null);
  const [video, setVideo] = useState(null);
  const [name, setName] = useState('');
  const [saved, setSaved] = useState(undefined);
  const [shareObj, setShareObj] = useState(undefined);

  const handleReset = useCallback(() => {
    setMode(null);
    setVideo(null);
    setName('');
    setSaved(undefined);
    setShareObj(undefined);
  }, [setMode, setVideo, setName, setSaved, setShareObj]);

  useEffect(() => {
    async function save() {
      const createdObj = await createSharedContent({ name, file: video });
      setSaved(true);
      setShareObj(createdObj);
    }

    if(mode && video && name && !saved) {
      setSaved(false);
      save();
    }
  }, [mode, video, name, saved, setSaved, setShareObj]);

  return (
    <>
      {!mode && <ModePicker onChange={(mode) => setMode(mode)} />}
      {mode === 'record' && <VideoRecorder onChange={(video) => setVideo(video)} />}
      {mode === 'upload' && <NotImplementedError onReset={handleReset} />}
      {video && <NameField onChange={(name) => setName(name)} />}
      {name && saved === false && <Loading>Uploading video, please wait...</Loading>}
      {name && saved === true && shareObj && <ShareOptions id={shareObj.id} onReset={handleReset} />}
    </>
  );
}

const ModePicker = ({ onChange }) => (
  <div className="mb-3">
    Do you want to
    {' '}
    <Anchor onClick={() => onChange('record')}>
      record a video
    </Anchor>
    {' '}
    or
    {' '}
    <Anchor onClick={() => onChange('upload')}>
      upload a file
    </Anchor>
    ?
  </div>
);

const VideoRecorder = ({ onChange }) => (
  <div className="w-full md:w-96 mb-3">
    <ReactVideoRecorder
      isOnInitially
      isFlipped={false}
      mimetype="video/webm"
      countdownTime={0}
      replayVideoAutoplayAndLoopOff
      disablePictureInPicture
      onRecordingComplete={videoBlob => onChange(videoBlob)}
    />
  </div>
);


const NameField = ({ onChange }) => {
  const [savedValue, setValue] = useState('');
  return (
    <div className="mb-3">
      Share with
      {' '}
      <input
        className="border border-gray-400 rounded shadow-md px-3 py-2"
        placeholder="Type the patient name"
        type="text"
        onChange={({ target: { value } }) => setValue(value)}
        onBlur={() => onChange(savedValue)}
        onKeyDown={({ key }) => (key === 'Enter') && onChange(savedValue)}
      />
    </div>
  );
};

const ShareOptions = ({ id, onReset }) => {
  const path = `/share/${id}`;
  const url = `${location.origin}${path}`;
  return (
    <div className="mb-3">
      <div className="mb-2">How to share?</div>
      <div className="flex justify-between items-center w-100">
        <ul className="flex flex-row">
          <li className="mr-2 p-2 bg-blue-300 hover:bg-blue-400">
            <a
              href={`whatsapp://send?text=Video shared through MediShare: ${url}`}
              data-action="share/whatsapp/share"
            >
              Whatsapp
            </a>
          </li>
          <li className="mr-2 p-2 bg-blue-300 hover:bg-blue-400">
            <CopyToClipboard
              text={url}
              label="Copy Url"
            />
          </li>
          <li className="mr-2 p-2 bg-blue-300 hover:bg-blue-400">
            <Link href={path}>
              <a>
                Visit
              </a>
            </Link>
          </li>
        </ul>
        <Anchor onClick={onReset}>
          Reset data
        </Anchor>
      </div>
    </div>
  );
}

const NotImplementedError = ({ onReset }) => (
  <div>
    <div>
      Not Implemented yet! Check it out later!
    </div>
    <div className="mt-3">
      <Anchor onClick={onReset}>
        Reset data
      </Anchor>
    </div>
  </div>
);

export default ShareForm;
