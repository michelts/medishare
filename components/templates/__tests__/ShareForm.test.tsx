import { render, screen, fireEvent } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import ReactVideoRecorder from 'react-video-recorder';
import { SharedContentFactory } from '@factories';
import createSharedContent from '@services/createSharedContent';
import ShareForm from '../ShareForm';

jest.mock('@services/createSharedContent', () => jest.fn());

jest.mock('react-video-recorder', () => jest.fn(({ onRecordingComplete }) => {
  // Trigger onRecordingComplete to simulate the user action
  setTimeout(() => onRecordingComplete('video-blob'), 10);
  return <div aria-label="Video Recorder" />;
}));

describe('ShareForm component', () => {
  const recipientName = 'John Smith';

  it('should create the share obj after the user record video, set dest and pick the media', async () => {
    const sharedContent = SharedContentFactory.build();
    createSharedContent.mockResolvedValue(sharedContent);

    render(<ShareForm />);
    fireEvent.click(screen.getByRole('button', { name: /record a video/ }));
    assertVideoRecorderWasRendered();

    const nameField = await screen.findByPlaceholderText(/patient name/);
    fireEvent.change(nameField, { target: { value: recipientName } });
    fireEvent.blur(nameField);
    expect(nameField).toHaveValue(recipientName);

    screen.getByText('Uploading video, please wait...');
    expect(createSharedContent).toHaveBeenCalledWith({ name: recipientName, file: 'video-blob' });

    await screen.findByRole('link', { name: 'Whatsapp' });
  });

  function assertVideoRecorderWasRendered() {
    fireEvent.click(screen.getByLabelText('Video Recorder'));
    expect(ReactVideoRecorder.mock.calls).toMatchSnapshot();
  }
});
