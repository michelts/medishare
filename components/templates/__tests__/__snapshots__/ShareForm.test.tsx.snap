// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareForm component should create the share obj after the user record video, set dest and allow share through whatsapp 1`] = `
Array [
  Array [
    Object {
      "disablePictureInPicture": true,
      "isFlipped": true,
      "isOnInitially": true,
      "mimetype": "video/webm",
      "onRecordingComplete": [Function],
      "replayVideoAutoplayAndLoopOff": true,
    },
    Object {},
  ],
]
`;
